Mathlib/Algebra/Homology/ShortComplex/QuasiIso.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2023 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+
+import Mathlib.Algebra.Homology.ShortComplex.Homology
+
+/-!
+# Quasi-isomorphisms of short complexes
+
+This file introduces the typeclass `QuasiIso φ` for a morphism `φ : S₁ ⟶ S₂`
+of short complexes (which have homology): the condition is that the induced
+morphism `homologyMap φ` in homology is an isomorphism.
+
+-/
+
+namespace CategoryTheory
+
+open Category Limits
+
+namespace ShortComplex
+
+variable {C : Type _} [Category C] [HasZeroMorphisms C]
+  {S₁ S₂ S₃ S₄ : ShortComplex C}
+  [S₁.HasHomology] [S₂.HasHomology] [S₃.HasHomology] [S₄.HasHomology]
+
+/-- A morphism `φ : S₁ ⟶ S₂` of short complexes that have homology is a quasi-isomorphism if
+the induced map `homologyMap φ : S₁.homology ⟶ S₂.homology` is an isomorphism. -/
+class QuasiIso (φ : S₁ ⟶ S₂) : Prop where
+  /-- the homology map is an isomorphism -/
+  isIso' : IsIso (homologyMap φ)
+
+instance QuasiIso.isIso (φ : S₁ ⟶ S₂) [QuasiIso φ] : IsIso (homologyMap φ) := QuasiIso.isIso'
+
+lemma quasiIso_iff (φ : S₁ ⟶ S₂) :
+    QuasiIso φ ↔ IsIso (homologyMap φ) := by
+  constructor
+  · intro h
+    infer_instance
+  · intro h
+    exact ⟨h⟩
+
+instance quasiIso_of_isIso (φ : S₁ ⟶ S₂) [IsIso φ] : QuasiIso φ :=
+  ⟨IsIso.of_iso (homologyMapIso (asIso φ))⟩
+
+lemma quasiIso_comp' (φ : S₁ ⟶ S₂) (φ' : S₂ ⟶ S₃) (hφ : QuasiIso φ) (hφ' : QuasiIso φ') :
+    QuasiIso (φ ≫ φ') := by
+  rw [quasiIso_iff] at hφ hφ' ⊢
+  rw [homologyMap_comp]
+  infer_instance
+
+instance quasiIso_comp (φ : S₁ ⟶ S₂) (φ' : S₂ ⟶ S₃) [QuasiIso φ] [QuasiIso φ'] :
+    QuasiIso (φ ≫ φ') :=
+  quasiIso_comp' φ φ' inferInstance inferInstance
```

Comment:
You are essentially duplicating every instance as a lemma. But since Lean 4 permits named variables, I wonder whether it is enough to just provide the instances and give nice names to the instance variables:
```suggestion
instance quasiIso_comp (φ : S₁ ⟶ S₂) (φ' : S₂ ⟶ S₃) [hφ : QuasiIso φ] [hφ' : QuasiIso φ'] :
    QuasiIso (φ ≫ φ') := by
  rw [quasiIso_iff] at hφ hφ' ⊢
  rw [homologyMap_comp]
  infer_instance
```
This can then be used as `quasiIso_comp φ φ' (hφ := foo) (hφ' := bar)` when needed.

Reply:
Thanks! I have struggled a little bit with the syntax to make this work downstream, but it works. (It seems the exact syntax would be `quasiIso_comp (hφ := foo) (hφ' := bar)` because from the type of `hφ`, Lean would find the type of `φ`!)